jni/lib/src/jni_class.dart

```diff
@@ -0,0 +1,160 @@
+  @pragma('vm:prefer-inline')
+  void _checkDeleted() {
+    if (_deleted) {
+      throw UseAfterFreeException(this, _cls);
+    }
+  }
+
+  JMethodID getConstructorID(String signature) {
+    _checkDeleted();
+    final methodSig = signature.toNativeChars();
+    final methodID = _env.GetMethodID(_cls, ctorLookupChars, methodSig);
+    _env.checkException();
+    calloc.free(methodSig);
+    return methodID;
+  }
+
+  /// Construct new object using [ctor].
+  JniObject newObject(JMethodID ctor, List<dynamic> args) {
+    _checkDeleted();
+    final jvArgs = JValueArgs(args, _env);
+    final newObj = _env.NewObjectA(_cls, ctor, jvArgs.values);
+    _env.checkException();
+    jvArgs.disposeIn(_env);
+    calloc.free(jvArgs.values);
+    return JniObject.of(_env, newObj, nullptr);
+  }
+
+  JMethodID _getMethodID(String name, String signature, bool isStatic) {
+    _checkDeleted();
+    final methodName = name.toNativeChars();
+    final methodSig = signature.toNativeChars();
+    final result = isStatic
+        ? _env.GetStaticMethodID(_cls, methodName, methodSig)
+        : _env.GetMethodID(_cls, methodName, methodSig);
+    _env.checkException();
+    calloc.free(methodName);
+    calloc.free(methodSig);
+    return result;
+  }
+
+  JFieldID _getFieldID(String name, String signature, bool isStatic) {
+    _checkDeleted();
+    final methodName = name.toNativeChars();
+    final methodSig = signature.toNativeChars();
+    final result = isStatic
+        ? _env.GetStaticFieldID(_cls, methodName, methodSig)
+        : _env.GetFieldID(_cls, methodName, methodSig);
+    _env.checkException();
+    calloc.free(methodName);
+    calloc.free(methodSig);
+    return result;
+  }
+
+  @pragma('vm:prefer-inline')
+  JMethodID getMethodID(String name, String signature) {
+    return _getMethodID(name, signature, false);
+  }
+
+  @pragma('vm:prefer-inline')
+  JMethodID getStaticMethodID(String name, String signature) {
+    return _getMethodID(name, signature, true);
+  }
+
+  @pragma('vm:prefer-inline')
+  JFieldID getFieldID(String name, String signature) {
+    return _getFieldID(name, signature, false);
+  }
+
+  @pragma('vm:prefer-inline')
+  JFieldID getStaticFieldID(String name, String signature) {
+    return _getFieldID(name, signature, true);
+  }
+
+  /// Returns the underlying [JClass].
+  JClass get jclass {
+    _checkDeleted();
+    return _cls;
+  }
+
+  JniGlobalClassRef getGlobalRef() {
+    _checkDeleted();
+    return JniGlobalClassRef._(_env.NewGlobalRef(_cls));
+  }
+
+  void delete() {
+    if (_deleted) {
+      throw DoubleFreeException(this, _cls);
+    }
+    _env.DeleteLocalRef(_cls);
+    _deleted = true;
+  }
+
+  /// Use this [JniClass] to execute callback, then delete.
+  ///
+  /// Useful in expression chains.
+  T use<T>(T Function(JniClass) callback) {
+    // TODO: Maybe use a mixin or something?
+    // JniClass and JniObject have some similar functionality.
+    _checkDeleted();
+    var result = callback(this);
```

Comment:
Might be good to catch any exceptions this throws, call `delete()`, then rethrow the exception?